docs/src/components/home/ElementPointer.tsx

```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import Box, { BoxProps } from '@material-ui/core/Box';
+import Typography from '@material-ui/core/Typography';
+import { debounce } from '@material-ui/core/utils';
+
+const PointerContext = React.createContext<undefined | React.Dispatch<React.SetStateAction<Data>>>(
+  undefined,
+);
+
+export const withPointer = <T extends React.ElementType>(
+  Component: T,
+  options: { id: string; name: string },
+) => {
+  const WithPointer = (props: unknown) => {
+    const root = React.useRef<null | HTMLElement>(null);
+    const setData = React.useContext(PointerContext);
+    return (
+      <React.Fragment>
+        {/* @ts-ignore */}
+        <Component
+          ref={root}
+          {...props}
+          onMouseOver={(event: React.MouseEvent) => {
+            event.stopPropagation();
+            if (setData && root.current) {
+              setData({
+                id: options.id,
+                target: root.current,
+                name: options.name,
+              });
+            }
+          }}
+        />
+      </React.Fragment>
+    );
+  };
+
+  return WithPointer as T;
+};
+
+export type Data = { id: null | string; name: null | string; target: null | HTMLElement };
+
+export default function PointerContainer({
+  onElementChange,
+  ...props
+}: BoxProps & { onElementChange?: (data: Data) => void }) {
+  const container = React.useRef<null | HTMLDivElement>(null);
+  const [data, setData] = React.useState<Data>({
+    id: null,
+    name: null,
+    target: null,
+  });
+  const debouncedSetData = React.useRef(debounce(setData, 200));
```

Comment:
@eps1lon The reason for debounce is to improve the ux when moving mouse between element. This demo is without debounce, https://codesandbox.io/s/element-pointer-j0p8q?file=/src/App.tsx

Reply:
Then debounce the specific call. Not all state updates. Especially not with a ref.

Reply:
> The reason for debounce is to improve the ux when moving mouse between element. 

I have never seen a debounce improve UX. It was always a crutch to fix performance bottlenecks.

Reply:
@eps1lon Is the demo in codesandbox does not bother you when moving mouse between elements? then compare with [debounce version](https://codesandbox.io/s/element-pointer-debounced-yty17?file=/src/App.tsx).